File: string_algorithms/week2_bwt_suffix_arrays/2_bwtinverse/bwtinverse.cpp
#include <algorithm>
#include <iostream>
#include <string>
#include <unordered_map>
#include <vector>

using std::string;
using std::unordered_map;
using std::vector;

const vector<char> LETTERS = {'$', 'A', 'C', 'G', 'T'};

vector<int> map_last_col_to_first_col(string bwt) {
  unordered_map<char, int> counts;
  for (char c : LETTERS)
    counts[c] = 0;
  for (char c : bwt)
    counts[c]++;

  int total_count = -1;
  unordered_map<char, int> position;
  for (char letter : LETTERS) {
    total_count += counts[letter];
    position[letter] = total_count;
  }

  vector<int> array(bwt.size(), 0);
  for (int i = bwt.size() - 1; i >= 0; i--) {
    array[i] = position[bwt[i]];
    position[bwt[i]]--;
  }

  return array;
}

string bwtinverse(string bwt) {
  vector<int> last_to_first = map_last_col_to_first_col(bwt);

  string result(bwt.size(), '$');
  int pos = 0;
  for (int i = 1; i < bwt.size(); i++) {
    result[bwt.size() - i - 1] = bwt[pos];
    pos = last_to_first[pos];
  }

  return result;
}

int main() {
  string bwt;
  std::cin >> bwt;
  std::cout << bwtinverse(bwt) << std::endl;
  return 0;
}

File: string_algorithms/week2_bwt_suffix_arrays/2_bwtinverse/bwtinverse.py
from typing import Final, List

LETTERS: Final = ["$", "A", "C", "G", "T"]


def map_last_col_to_first_col(bwt: str) -> List[int]:  # noqa: C901
    counts = {char: 0 for char in LETTERS}
    for char in bwt:
        counts[char] += 1

    total_count = -1
    position = {}
    for letter in LETTERS:
        total_count += counts[letter]
        position[letter] = total_count

    array = [0] * len(bwt)
    for i in range(len(bwt) - 1, -1, -1):
        array[i] = position[bwt[i]]
        position[bwt[i]] -= 1

    return array


def bwtinverse(bwt: str) -> str:
    last_to_first = map_last_col_to_first_col(bwt)

    result = ["$"] * len(bwt)
    pos = 0
    for i in range(1, len(bwt)):
        result[-(i + 1)] = bwt[pos]
        pos = last_to_first[pos]

    return "".join(result)


if __name__ == "__main__":
    bwt = input()
    print(bwtinverse(bwt))
